aleph-client/src/pallets/multisig.rs

```diff
@@ -71,3 +129,388 @@ impl<S: SignedConnectionApi> MultisigUserApi for S {
+
+        ensure!(
+            sorted_signatories.len() > 1,
+            "There must be at least 2 different signatories"
+        );
+        ensure!(
+            sorted_signatories.len() >= threshold as usize,
+            "Threshold must not be greater than the number of unique signatories"
+        );
+        ensure!(
+            threshold >= 2,
+            "Threshold must be at least 2 - for threshold 1, use `as_multi_threshold_1`"
+        );
+
+        Ok(Self {
+            signatories: sorted_signatories,
+            threshold,
+        })
+    }
+
+    /// The multisig account derived from signatories and threshold.
+    ///
+    /// This method is copied from the pallet, because:
+    ///  -  we don't want to add a new dependency
+    ///  -  we cannot instantiate pallet object here anyway (the corresponding functionality exists
+    ///     as pallet's method rather than standalone function)
+    pub fn account(&self) -> AccountId {
+        let entropy =
+            (b"modlpy/utilisuba", &self.signatories, &self.threshold).using_encoded(blake2_256);
+        Decode::decode(&mut TrailingZeroInput::new(entropy.as_ref()))
+            .expect("infinite length input; no invalid inputs for type; qed")
+    }
+}
+
+/// A context in which ongoing signature aggregation is performed.
+#[derive(Clone, Eq, PartialEq, Debug)]
+pub struct Context {
+    /// The entity for which aggregation is being made.
+    party: MultisigParty,
+    /// Derived multisig account (the source of the target call).
+    author: AccountId,
+
+    /// Pallet's coordinate for this aggregation.
+    timepoint: Timepoint,
+    /// Weight limit when dispatching the call.
+    max_weight: Weight,
+
+    /// The target dispatchable, if already provided.
+    call: Option<Call>,
+    /// The hash of the target dispatchable.
+    call_hash: CallHash,
+
+    /// The set of accounts, that already approved the call (via this context object), including the
+    /// author.
+    ///
+    /// `approvers.len() < party.threshold` always holds.
+    approvers: HashSet<AccountId>,
+}
+
+impl Context {
+    /// In case `Context` object has been passed somewhere, where this limit should be adjusted, we
+    /// allow for that.
+    ///
+    /// Actually, this isn't used until threshold is met, so such changing is perfectly safe.
+    pub fn change_max_weight(&mut self, max_weight: Weight) {
+        self.max_weight = max_weight;
+    }
+
+    /// Set `call` only if `self.call_hash` is matching.
+    fn set_call(&mut self, call: &Call) -> anyhow::Result<()> {
+        ensure!(
+            self.call_hash == compute_call_hash(call),
+            "Call doesn't match to the registered hash"
+        );
+        self.call = Some(call.clone());
+        Ok(())
+    }
+
+    /// Register another approval. Consume `self` if the threshold has been met and `call` is
+    /// already known.
+    fn add_approval(mut self, approver: AccountId) -> Option<Self> {
```

Comment:
there is no way to determine who approved, who was author etc after this consumes self which might be useful information loss.
suggestion: `fn add_approval(...) -> Either<Self, ConsumedContext>` where `ConsumedContext` only allows to read fields or smt like this

Reply:
done